### PR TITLE
[API] dont reqToken on GetReactions (fix #9543)

### LINKED ISSUE #9543: [Feature Request] Allow GET comments reactions without token via API
## Description

Related to #9220

Currently, we can:

- GET issues: https://gitea.com/api/v1/repos/meteorlxy/vssue-test/issues?q=[Vssue]Vssue+Dev
- GET comments: https://gitea.com/api/v1/repos/meteorlxy/vssue-test/issues/1/comments

But we can't GET reactions if we don't pass token (get 401):

- GET reactions: https://gitea.com/api/v1/repos/meteorlxy/vssue-test/issues/comments/104283/reactions

Hope the behavior of reactions API can be consistent with the former two APIs.

cc @6543

### PR DESCRIPTION
**Pull request by 6543**
fix #9543

### CODE REVIEW

**Comment by zeripath** [comment]
I think you need to take a look at:

https://github.com/go-gitea/gitea/blob/d59536c222857713d95846d77903282afe608f23/routers/api/v1/repo/issue_reaction.go#L58

Looking at:

https://github.com/go-gitea/gitea/blob/d59536c222857713d95846d77903282afe608f23/routers/api/v1/api.go#L712

I don't think you need the check in issue_reaction.go - and if you're anonymous then I suspect that `ctx.User.IsAdmin` might cause a nil pointer.

**Comment by 6543** [comment]
@zeripath to https://github.com/go-gitea/gitea/pull/9548/files#diff-3afa9b9bab64821b908bf6a97aa8e2d2R61

does Go exed secod statement if first is false -> in other languages this isn't the case

**Comment by 6543** [comment]
@zeripath done

File: routers/api/v1/repo/issue_reaction.go [hunk 1]
```diff
@@ -55,9 +57,20 @@ func GetIssueCommentReactions(ctx *context.APIContext) {
 		return
 	}
 
-	if !ctx.Repo.CanRead(models.UnitTypeIssues) && !ctx.User.IsAdmin {
-		ctx.Error(http.StatusForbidden, "GetIssueCommentReactions", errors.New("no permission to get reactions"))
-		return
+	if ctx.Repo.Repository.IsPrivate {
```
Comment (lafriks) [inline routers/api/v1/repo/issue_reaction.go]: Shouldn't we also check if organization is private?

Comment (6543) [reply routers/api/v1/repo/issue_reaction.go]: @lafriks can a repo be public when a org is private ?!?

Comment (lafriks) [reply routers/api/v1/repo/issue_reaction.go]: But I don't think it will be visible then, but that needs to be verified

Comment (6543) [reply routers/api/v1/repo/issue_reaction.go]: On UI I have no acces (404) on souch a repo (loged in/loged out)

Comment (6543) [reply routers/api/v1/repo/issue_reaction.go]: and then access permission is checed below (L66 R)

Comment (lafriks) [reply routers/api/v1/repo/issue_reaction.go]: Yes but if it will be public and org private it will not get to line 66

**Comment by 6543** [comment]
@lafriks does this https://github.com/go-gitea/gitea/pull/9548/files#diff-029cb1fefb1ae53a0ef72d3fd38a7f7dR16 resolfe your permission issue ?

-> I created a new file to make it reusable ...

File: routers/api/v1/utils/permission.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/context"
+	api "code.gitea.io/gitea/modules/structs"
+)
+
+func HasUserAccessToIssues(ctx *context.APIContext) bool {
+	hiddenOrg := false
+	if ctx.Repo.Owner.IsOrganization() {
+		if ctx.User == nil && (ctx.Repo.Owner.Visibility == api.VisibleTypePrivate || ctx.Repo.Owner.Visibility == api.VisibleTypeLimited) {
+			return false
```
Comment (lafriks) [inline routers/api/v1/utils/permission.go]: this should probably be:
```
ctx.NotFound()
return false
```

Comment (6543) [reply routers/api/v1/utils/permission.go]: @lafriks it is

File: routers/api/v1/repo/issue_reaction.go [hunk 3]
```diff
@@ -258,8 +262,7 @@ func GetIssueReactions(ctx *context.APIContext) {
 		return
 	}
 
-	if !ctx.Repo.CanRead(models.UnitTypeIssues) && !ctx.User.IsAdmin {
-		ctx.Error(http.StatusForbidden, "GetIssueReactions", errors.New("no permission to get reactions"))
+	if !utils.HasUserAccessToIssues(ctx) {
```
Comment (zeripath) [inline routers/api/v1/repo/issue_reaction.go]: Hasn't this test been checked by

https://github.com/6543-forks/gitea/blob/a32425f070c67d56a894b1573f95f3e8ee9b17f8/routers/api/v1/api.go#L712

Comment (6543) [reply routers/api/v1/repo/issue_reaction.go]: `mustEnableIssuesOrPulls`:

https://github.com/6543-forks/gitea/blob/a32425f070c67d56a894b1573f95f3e8ee9b17f8/routers/api/v1/api.go#L456

only check if Issues are enabled - as far as i can tell

Comment (zeripath) [reply routers/api/v1/repo/issue_reaction.go]: Hmm but what about the issues endpoints above these changed reaction endpoints?

Comment (6543) [reply routers/api/v1/repo/issue_reaction.go]: no endpoints changed ?!? - just allow noneLogedin API requests to public repos @zeripath

Comment (zeripath) [reply routers/api/v1/repo/issue_reaction.go]: Remember this is checked too:

https://github.com/go-gitea/gitea/blob/aee1133e505a26a81ffad32c029ea093dc8feb92/routers/api/v1/api.go#L804

Either the reactions need these additional checks - in which case so do the issue comments endpoints, or this is just repeating work.

The issue comments and their reactions should have the same the permissions checks.

File: routers/api/v1/repo/issue_reaction.go [hunk 4]
```diff
@@ -55,8 +58,7 @@ func GetIssueCommentReactions(ctx *context.APIContext) {
 		return
 	}
 
-	if !ctx.Repo.CanRead(models.UnitTypeIssues) && !ctx.User.IsAdmin {
-		ctx.Error(http.StatusForbidden, "GetIssueCommentReactions", errors.New("no permission to get reactions"))
+	if !utils.HasUserAccessToIssues(ctx) {
```
Comment (lunny) [inline routers/api/v1/repo/issue_reaction.go]: Why not just `if !ctx.Repo.CanRead(models.UnitTypeIssues)`

**Comment by 6543** [comment]
@zeripath I used lunny's suggestion ... I did'nt know there is already souch a function :+1:
@lunny done

**Comment by 6543** [comment]
@zeripath @lunny I'll leave it as it is for now ... without invent rights check for one endpoint

but there is a securrey issue in **general** !! -> PM

**Comment by 6543** [comment]
@zeripath `!ctx.Repo.CanRead(models.UnitTypeIssues) && !ctx.User.IsAdmin` works without login -> if issue is public, it works if not 404 is returned

**Comment by lunny** [comment]
I think we should keep `!ctx.Repo.CanRead(models.UnitTypeIssues)` and remove `!ctx.User.IsAdmin`, because the previous has included the next. And `ctx.User` may be `nil`.

**Comment by 6543** [comment]
@lunny is this what you mean? fdf7965d31b84d7287f06c22f818206e52d30e87

**Review by lunny — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
ready to merge :rocket:

**Pull request merged**